pkg/filesystem/subvolume.go

```diff
@@ -79,28 +85,57 @@ func listCephFSSubvolumes(ctx context.Context, clientsets *k8sutil.Clientsets, o
 			}
 			// append the subvolume which doesn't have any snapshot attached to it.
 			for _, sv := range subvol {
+				state := getSubvolumeState(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name, sv.Name, svg.Name)
+
 				// Assume the volume is stale unless proven otherwise
 				stale := true
 				// lookup for subvolume in list of the PV references
 				_, ok := subvolumeNames[sv.Name]
-				if ok || checkSnapshot(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name, sv.Name, svg.Name) {
-					// The volume is not stale if a PV was found, or it has a snapshot
+				if ok {
+					// The volume is not stale if a PV was found
 					stale = false
 				}
-				status := "stale"
+				status := Stale
 				if !stale {
 					if includeStaleOnly {
 						continue
 					}
-					status = "in-use"
+					status = InUse
+				} else {
+					// check the state of the stale subvolume
+					// if it is snapshot-retained then skip listing it.
+					if state == "snapshot-retained" {
+						status = state
+						continue
+					}
+					// check if the stale subvolume has snapshots.
+					if checkSnapshot(ctx, clientsets, operatorNamespace, clusterNamespace, fs.Name, sv.Name, svg.Name) {
+						status = StaleWithSnapshot
+					}
+
 				}
-				subvolumeNames[sv.Name] = subVolumeInfo{fs.Name, svg.Name}
+				subvolumeNames[sv.Name] = subVolumeInfo{fs.Name, svg.Name, state}
 				fmt.Println(fs.Name, sv.Name, svg.Name, status)
```

Comment:
Println is not a formatted output when we have large number of subvolumes it difficult to read, can you please open an issue to print as formatted like `kubectl output`

Reply:
https://github.com/rook/kubectl-rook-ceph/issues/211 , it is already added here, I will check on how we can have a proper format for it.